Doc/tools/extensions/grammar_snippet.py

```diff
@@ -0,0 +1,192 @@
+    # To get around this, we set it to this non-empty string:
+    rawsource = 'You should not see this.'
+
+    literal = nodes.literal_block(
+        rawsource,
+        '',
+        # TODO: Use a dedicated CSS class here and for strings.
+        # and add it to the theme too
+        classes=['highlight'],
+    )
+
+    grammar_re = re.compile(
+        """
+            (?P<rule_name>^[a-zA-Z0-9_]+)     # identifier at start of line
+            (?=:)                             # ... followed by a colon
+        |
+            [`](?P<rule_ref>[a-zA-Z0-9_]+)[`] # identifier in backquotes
+        |
+            (?P<single_quoted>'[^']*')        # string in 'quotes'
+        |
+            (?P<double_quoted>"[^"]*")        # string in "quotes"
+        """,
+        re.VERBOSE,
+    )
+
+    for line in content:
+        last_pos = 0
+        for match in grammar_re.finditer(line):
+            # Handle text between matches
+            if match.start() > last_pos:
+                literal += nodes.Text(line[last_pos : match.start()])
+            last_pos = match.end()
+
+            # Handle matches
+            groupdict = {
+                name: content
+                for name, content in match.groupdict().items()
+                if content is not None
+            }
+            match groupdict:
+                case {'rule_name': name}:
+                    name_node = addnodes.literal_strong()
+
+                    # Cargo-culted magic to make `name_node` a link target
+                    # similar to Sphinx `production`.
+                    # This needs to be the same as what Sphinx does
+                    # to avoid breaking existing links.
+                    domain = directive.env.domains['std']
+                    obj_name = f"{group_name}:{name}"
+                    prefix = f'grammar-token-{group_name}'
+                    node_id = make_id(
+                        directive.env, directive.state.document, prefix, name
+                    )
+                    name_node['ids'].append(node_id)
+                    directive.state.document.note_implicit_target(
+                        name_node, name_node
+                    )
+                    domain.note_object(
+                        'token', obj_name, node_id, location=name_node
+                    )
+
+                    text_node = nodes.Text(name)
+                    name_node += text_node
+                    literal += name_node
+                case {'rule_ref': name}:
+                    ref_node = addnodes.pending_xref(
+                        name,
+                        reftype="token",
+                        refdomain="std",
+                        reftarget=f"{group_name}:{name}",
+                    )
+                    ref_node += nodes.Text(name)
+                    literal += ref_node
+                case {'single_quoted': name} | {'double_quoted': name}:
+                    string_node = nodes.inline(classes=['nb'])
```

Comment:
I assume `'nb'` is a Pygments class? Is there a way to avoid hard-coding this?

Reply:
Yes, it's a Pygments class.
This is a chicken-and-egg problem. My plan is to merge `nb`, then add a new class to python-docs-theme, and then go back here to use the new class.

Reply:
OK. Another option is adding a small CSS file in `Doc/tools/static/` and using `app.add_css_file()` which we would then later merge in to p-d-t.

Reply:
Or add it to p-d-t now and we can release it right away?

Reply:
I'll add the Pygments class for "other strings".